torch/csrc/api/include/torch/data/dataloader2.h

```diff
@@ -0,0 +1,247 @@
+  }
+
+  /// Joins the `DataLoader2`'s worker threads and drains internal queues.
+  /// This function may only be invoked from the main thread (in which the
+  /// `DataLoader2` lives).
+  void join() {
+    if (joined_) {
+      return;
+    }
+    shuttle_.drain();
+    // Send one 'quit' message per worker. Since a worker dies (exits its
+    // thread) after receiving this message, each `QuitWorker()` message will be
+    // read by exactly one worker.
+    for (size_t w = 0; w < options_.workers; ++w) {
+      push_job(QuitWorker());
+    }
+    for (auto& worker : workers_) {
+      worker.join();
+    }
+    joined_ = true;
+  }
+
+  /// Returns the options with which the `DataLoader2` was configured.
+  const FullDataLoaderOptions& options() const noexcept {
+    return options_;
+  }
+
+ private:
+  /// Simple mix-in to give something a sequence number.
+  struct Sequenced {
+    Sequenced() = default;
+    Sequenced(size_t sqn) : sequence_number(sqn) {}
+    size_t sequence_number;
+  };
+
+  struct QuitWorker {};
+  struct StopEpoch {};
+
+  /// A `Job` is either a `BatchRequest` (new indices to fetch data at) or a
+  /// `QuitWorker` object, to indicate the worker should shut down.
+  struct Job : Sequenced {
+    Job() = default;
+    Job(QuitWorker q, size_t sqn) : Sequenced(sqn), quit(q) {}
+    Job(BatchRequest&& i, size_t sqn)
+        : Sequenced(sqn), batch_request(std::move(i)) {}
+    optional<QuitWorker> quit;
+    optional<BatchRequest> batch_request;
+  };
+
+  /// The finished result of a job.
+  struct Result : Sequenced {
+    Result() = default;
+    Result(optional<UnwrappedBatch>&& b, size_t sqn)
+        : Sequenced(sqn), batch(std::move(b)) {}
+    Result(StopEpoch s, size_t sqn) : Sequenced(sqn), stop_epoch(s) {}
+    Result(std::exception_ptr exception, size_t sqn)
+        : Sequenced(sqn), exception(std::move(exception)) {}
+    optional<UnwrappedBatch> batch;
+    optional<StopEpoch> stop_epoch;
+    std::exception_ptr exception;
+  };
+
+  /// Resets the internal state of the `DataLoader2`, optionally pre-fetching
+  /// new jobs.
+  void reset(bool prefetch = true) {
+    shuttle_.drain();
+    sequence_number_ = 0;
+    sequencer_ = new_sequencer();
+    dataset_.reset();
+    if (prefetch && options_.workers > 0) {
+      this->prefetch();
+    }
+  }
+
+  /// Schedules `requested_jobs` many new batches to be fetched. The actual
+  /// number of jobs scheduled may be less if the `DataLoader2` exhausts.
+  void prefetch(size_t requested_jobs) {
+    while (requested_jobs-- > 0) {
+      push_job(options_.batch_size);
+    }
+  }
+
+  /// Schedules the maximum number of jobs (based on the `max_jobs` option).
+  void prefetch() {
+    prefetch(options_.max_jobs);
+  }
+
+  /// Returns the next batch of data, or an empty `optional` if the
+  /// `DataLoader2` is exhausted. This operation will block until a batch is
+  /// available.
+
+  optional<UnwrappedBatch> next() {
```

Comment:
This means, the application need to handle nullopt when iterating over a dataset. 
Say we have 4 dataloader threads and they all made requests to dataset in prefetch. One of them has data and others don't but the order is reversed. So the application will get three nullopts before getting the final batch. Instead, could we make next() always return non empty batches?

Reply:
No, the `Iterator` calls `next()` to get the next batch, and if it's empty it means `begin() == end()` and the epoch is over. You can see in the tests that the end user gets `Batch` and not `optional<Batch>`: https://github.com/pytorch/pytorch/pull/15096/files#diff-46d6ad7987c3572d26ce065f4986e4abR1153

Reply:
> We simply need to discard these empty batches

That's what happens here: https://github.com/pytorch/pytorch/pull/15096/files#diff-731545db723458d2f200428b7c3d563aR170

In this loop:

```
      while (auto result = pop_result()) {
        if (result->exception) {
          throw WorkerException(result->exception);


// If result->batch is empty optional, it is skipped and we go to the next iteration of this while loop
        } else if (result->batch) {
          prefetch(1);
          return std::move(result->batch);
        }
      }
```

Reply:
Ah, sorry for the confusion, this should be a trivial thing.